### PR TITLE
R-like extraction for tensors

### PR DESCRIPTION
**Pull request by goldingn**
This implements a (slightly cleaned up) port of greta's extraction behaviour as a solution to #198.

It should behave *exactly* like extraction on an R array, such that if a user creates an array `x` and extracts elements to an array `y` like this:
```r
y <- x[<some extraction call>]
```
then this should have an identical result:
```r
x_tf <- tf$constant(x)
y_tf <- x_tf[<some extraction call>]
y <- tf$Session()$run(y_tf)
```

There are tests to this effect, for all the extraction syntax that springs to mind. 

This retains 0-based indexing as an option, as in #195, but under a different name to make clear it uses the new default is complete R-like extraction syntax, not just 1-based indexing.

### CODE REVIEW

**Comment by goldingn** [comment]
I haven't attempted to spot 0-based indices here, because 0s are allowed in indices for R's array extraction syntax (those elements are just ignored):

```r
x <- array(1:3)
x[0:2]
```
```
[1] 1 2
```
and in the new TF implementation
```r
x_tf <- tf$constant(x)
y_tf <- x_tf[0:2]
tf$Session()$run(y_tf)
```
```
[1] 1 2
```

**Comment by jjallaire** [comment]
This looks awesome, definitely the way to go! Very much appreciated.

I realize your time is tight and I'm the deadline motivated one so I am happy to make any further tweaks discussed here after the PR is merged to master.

I realize that the 0-based syntax is valid in R but it seems mostly unnecessary (since it's ignored) so I'm wondering if emitting a warning for some period of time (e.g. a year) might be valuable? We could have this warning NOT fire if the user explicitly sets `options(tensorflow.r_like_extract = TRUE)` as opposed to receiving the default value.

**Comment by goldingn** [comment]
Yeah, a warning like that sounds sensible!

**Comment by jjallaire** [comment]
Is there a way to do that in your greta-implementation? (it looks like the indexes might be somewhat opaque). If there is any way whatsoever to do it I think we should as it will save users lots of time and trouble). I am happy to implement this if you should give me a pointer as to how.

**Comment by goldingn** [comment]
I imagine the best bet would be to crib code from the 0-indexing function - that grabs indices for each dimension which could be checked for 0s.

That code is super clunky and opaque because of having to parse missing arguments in the dots, but for a first pass you could just copy-paste all the code up until the strided_slice call, then lool for 0s in those indices. We can always tidy it up later.

**Comment by jjallaire** [comment]
@goldingn I'm noticing that some Keras tensor indexing code that used to work correctly is now failing with the revised implementation:

```r
library(keras)

model <- keras_model_sequential() %>% 
  layer_dense(units = 64, input_shape = c(784)) %>% 
  layer_dense(units = 10, activation = 'softmax')

model$output[,1]
```

Gives:

```
Error in prod(dims) : invalid 'type' (list) of argument
```

Whereas this used to give:

```
Tensor("StridedSlice:0", shape=(?,), dtype=float32)
```

So there is a failure mode we may not have thought about here. If this is too intracted perhaps we should consider modifying the old code to pass the tests and then migrating the greta code only when we have more time. Let me know what you think.

**Comment by goldingn** [comment]
I won't be able to get to a laptop for a bit to check this out, but my immediate guess would be that whatever class `model$output` is, its `dim()` method is behaving differently to simple tensors.

That could be cleared up either in keras, or by doing `dims_in <- unlist(dim(x))` at the top of `extract_r_like()`

**Comment by jjallaire** [comment]
The class is no different that a regular TF tensor:

```r
> class(model$output)
[1] "tensorflow.python.framework.ops.Tensor"      "tensorflow.python.framework.ops._TensorLike"
[3] "python.builtin.object"    
```

```r
> class(tf$constant(42))
[1] "tensorflow.python.framework.ops.Tensor"      "tensorflow.python.framework.ops._TensorLike"
[3] "python.builtin.object"    
```

**Comment by goldingn** [comment]
Huh, OK thanks for checking that out!
I'll take a look in the morning.

**Comment by jjallaire** [comment]
Okay, thanks! 

I've further narrowed the issue to the handling of placeholder tensors. I can also reproduce the error with:

```r
library(tensorflow)
x <- tf$placeholder(tf$float32, shape = shape(NULL, 100))
x[,1]
```
`Error in prod(dims) : invalid 'type' (list) of argument`

I think this is likely to the fact that the placeholder dimension is `NULL`:

```
> dim(x)
[[1]]
NULL

[[2]]
[1] 100
```

Which results in an error when this expression is evaluated:

```
> prod(list(NULL, 100))
Error in prod(list(NULL, 100)) : invalid 'type' (list) of argument
```

**Comment by jjallaire** [comment]
The simple error with the `tf$placeholder` is fixed by this commit: https://github.com/rstudio/tensorflow/pull/200/commits/2dfe92910a66720cafc6bcbb42a9d498ba9b2070

Which basically just substitutes 1 for any dimensions that are `NULL`

Unfortunately the Keras model code now breaks with:

```
 Error in `[[<-`(`*tmp*`, call_list[[1]], value = dummy_in) : 
  invalid subscript type 'language' 
```

**Comment by jjallaire** [comment]
A great clue, this code actually works:

```r
output <- model$output
output[,1]
```

So a simplier repro that fails would be:

```r
library(tensorflow)
container <- list(x = tf$constant(matrix(c(1:4), ncol = 2, nrow = 2)))
container$x[,1]
```

So the greta style indexing isn't handling list elements. Hopefully there is a simple fix for this!

**Comment by jjallaire** [comment]
In terms of cribbing code from the 0 indexing function, it looks like I would just need to crib code until `begin` is defined rather than all the way up the strided slice call, is that correct?

**Comment by goldingn** [comment]
I had forgotten about placeholders with unknown dimensions. That's incompatible with the strategy for full R-like extraction, but was already implemented in the 0-based extraction.

Note that setting unknown dimensions to 1 clears that particular error, but returns the wrong result. It should have shape `(?, )` or `(?, 1)` with `drop=TRUE`.

In #202 I've fixed this by using 'manual' 1-based extraction if there are unknown dimensions, and full R-like extraction otherwise.

I've also fixed the invalid subscript type error, and had a ho at detecting 0s in the indices

**Comment by dfalbel** [comment]
@jjallaire @goldingn 
It seems that this added a lot of overhead in extraction. See this benchmark for example:

```
> options(tensorflow.r_like_extract = FALSE)
> library(tensorflow)
> 
> x <- tf$placeholder(dtype = tf$float16, shape = c(3, 25, 71, 512))
> 
> microbenchmark::microbenchmark(
+   x[1,,,], times = 10
+ )
Unit: milliseconds
       expr      min       lq     mean   median       uq     max neval
 x[1, , , ] 2.496407 2.542692 5.116753 2.626734 3.194224 26.3578    10

Restarting R session...

> library(tensorflow)
> 
> x <- tf$placeholder(dtype = tf$float16, shape = c(3, 25, 71, 512))
> 
> microbenchmark::microbenchmark(
+   x[1,,,], times = 10
+ )
Unit: milliseconds
       expr      min       lq     mean   median       uq      max neval
 x[1, , , ] 894.8528 918.0925 940.6862 924.0265 936.2776 1025.532    10
```

**Comment by jjallaire** [comment]
Wow, that is a huge difference indeed.

I wonder if should consider implementing 1-based indexing for the `tf$strided_slice` implementation after all? I realize it's a bit more work for us but it seems like it could save users quite a bit of CPU time!

Here's the implementation in progress for that: https://github.com/rstudio/tensorflow/pull/198

It has 3 tests failing and I'm not quite sure how to resolve them. If we can get those tests to pass I could move the warning code over from the existing implementation.

@goldingn Interested in your take on whether this performance difference is actually worth living with for the additional syntactic flexibility. At initial glance it seems to me like 1 second is too long for a straightforward slicing operation.

**Comment by goldingn** [comment]
Yeesh, that is slow. 
I can do some profiling to see what the time suck is, though I suspect it's all the `tf$reshape()`ing. 
I'm actually more concerned about any slow down when executing the graph, rather than constructing it - we should definitely benchmark that.

Assuming the slow down is the `tf$reshape()`ing, the easiest fix is to drop full R-like extracting from the API, and use the 'manual' `tf$strided_slice()` one-based indexing that's already implemented for tensors with unknown dimensions. It should work fine when all dimensions are known too.

The `[` method in `generics.R` uses conditionals to select between the three extraction methods (zero-based manual, one-based manual, or R like). So it would be easy to change that and delete the unnecessary tests.

I'd be happy to do this, but won't have time until next week at the earliest, so very happy for someone else to implement instead if this is urgent!

**Comment by goldingn** [comment]
I think speed is probably more important than flexibility here - people using tensorflow will know that they'll need to use a slightly different API than for arrays.

That's slightly different for greta, where identical behaviour to arrays is important. Over there I'll try to detect all-numeric, positive, increasing indices, and use the manual extraction in that case.

**Comment by jjallaire** [comment]
PR here: https://github.com/rstudio/tensorflow/pull/203
